## This is the implemantation of PCA

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns


Now create a dummy dataset

In [2]:
data=np.array([[3,7],[-4,-6],[2,4],[1,3],[5,6]])

In [3]:
dataframe=pd.DataFrame(data,columns=['Feature1','Feature2'])
dataframe

Feature1  Feature2
0         3         7
1        -4        -6
2         2         4
3         1         3
4         5         6

### Step 1- Data must be in Std. Normal form

Now , check if the data is in standard normal form or not
Mean=`0`
Standard Deviation=`1`

In [4]:
dataframe.describe()

Feature1  Feature2
count  5.000000  5.000000
mean   1.400000  2.800000
std    3.361547  5.167204
min   -4.000000 -6.000000
25%    1.000000  3.000000
50%    2.000000  4.000000
75%    3.000000  6.000000
max    5.000000  7.000000

we can see that the data is not in standard normal form by mean and standard deviation. so we have to standardize the data

In [5]:
from sklearn.preprocessing import StandardScaler

scaler=StandardScaler() #call the model
dataframe=scaler.fit_transform(dataframe)
dataframe

array([[ 0.53215208,  0.90875888],
       [-1.79601328, -1.90406622],
       [ 0.19955703,  0.25964539],
       [-0.13303802,  0.04327423],
       [ 1.19734219,  0.69238772]])

In [6]:
dataframe=pd.DataFrame(dataframe,columns=['Feature1','Feature2'])
dataframe

Feature1  Feature2
0  0.532152  0.908759
1 -1.796013 -1.904066
2  0.199557  0.259645
3 -0.133038  0.043274
4  1.197342  0.692388

In [7]:
dataframe.describe()

Feature1      Feature2
count  5.000000  5.000000e+00
mean   0.000000 -2.220446e-17
std    1.118034  1.118034e+00
min   -1.796013 -1.904066e+00
25%   -0.133038  4.327423e-02
50%    0.199557  2.596454e-01
75%    0.532152  6.923877e-01
max    1.197342  9.087589e-01

Now our dataframe is in standard normal form.
mean is almost =0
standard deviation is almost =1

Step 1-complete 
---
Step 2-calculate the covariance matrix
---

matrix can be calculated in two ways
1. by formula - manual calculation
2. by numpy

In [8]:
covariance_matrix=(dataframe.T@dataframe)/4
covariance_matrix

Feature1  Feature2
Feature1  1.250000  1.194602
Feature2  1.194602  1.250000

In [9]:
covariance_matrix1=np.cov(dataframe.T)
covariance_matrix1

array([[1.25      , 1.19460203],
       [1.19460203, 1.25      ]])

---
Step 3 - evaluate eigen value and eigen vectors using covariance matrix 
---

In [10]:
eigenValues,eigenVectors=np.linalg.eig(covariance_matrix)
print(eigenValues)
print(eigenVectors)

[0.05539797 2.44460203]
[[-0.70710678 -0.70710678]
 [ 0.70710678 -0.70710678]]


Highest Eigen Value=`2.44460203`

---
Step 4 - extract the principal component 1 
---

In [11]:
pc1=dataframe@eigenVectors[:,0]
pc1

0    0.266301
1   -0.076405
2    0.042489
3    0.124672
4   -0.357057
dtype: float64

Thus, instad of 2 features we have 1 principal component which can be used in place of those 2 features

- Since we wanted to reduce the dimensionality from 2 to 1, we can use the first principal component
but just to show , we can calculate the second principal component also

In [12]:
pc2=dataframe@eigenVectors[:,1]
pc2

0   -1.018878
1    2.616351
2   -0.324705
3    0.063473
4   -1.336241
dtype: float64

## PCA implementation using Library remaining 